### ISSUE TITLE
Won't display images served over HTTPS

### ISSUE THREAD

**Issue by ChristianTucker** [kfiroo/react-native-cached-image#10]
The native `<Image />` component has no problem serving over HTTPS, however this component is for some reason not loading anything over HTTPS. Changing the link to an HTTP link works fine. However, my CDN requires a HTTPS connection. 

For example, the following will not load

```//js
<CachedImage source={{uri: 'https://firebasestorage.googleapis.com/v0/b/qualify-college-dating.appspot.com/o/user_photos%2Fc9bd1771-e08c-4fc4-986d-7658776e6491?alt=media&token=9dae0c20-ef0a-40e9-ad87-25a8b8460f5c'}} style={{width: 300, height: 300}} />
```

However, the link in the browser loads the image fine, but something like

```//js
<CachedImage source={{uri: 'http://placehold.it/256'}} style={{width: 300, height: 300}} />
```

Loads just fine, also the native component loads just fine 

```//js
<Image source={{uri: 'https://firebasestorage.googleapis.com/v0/b/qualify-college-dating.appspot.com/o/user_photos%2Fc9bd1771-e08c-4fc4-986d-7658776e6491?alt=media&token=9dae0c20-ef0a-40e9-ad87-25a8b8460f5c'}} style={{width: 300, height: 300}} />
```

I'm completely uninstalled and re-installed the application and even did some debugging by creating a fork, which shows that the image is being downloaded and saved to disk. The image on the disk is being found during the `getCachedImagePath` call as well.  The path being returned in the `CachedImage.js` render() call is as follows:

```
Source: { "uri": "file:///Users/christiantucker/Library/Developer/CoreSimulator/Devices/4D19844A-CAEA-4CCA-A3C6-328C58CC6AFA/data/Containers/Data/Application/6C8A618F-0ED7-4F69-8CBE-A80F5DFA2CB1/Documents/firebasestorage.googleapis.com/4c79bcb33ecdfe4512febf4c97c40cb86082398d.com/o/user_photos%2Fc9bd1771-e08c-4fc4-986d-7658776e6491" }
```

This problem persists with any HTTPS link, even the ones in your example. All HTTP links are fine.

----

This may actually be a problem with react-native-fs as I've created my own caching implementation in the last few hours and I have the same problem with HTTPS.

**Comment by kfiroo** [comment]
@ChristianTucker Hey, my app uses only `https` for all assets and the component works fine, not sure why it would break for you :(

I'll dig around to try and find something, any information you can give me would help :)

What versions are you using? (react-native / react-native-fs / react-native-cached-image / any others that might be relevant?)

**Comment by ChristianTucker** [comment]
```
    "react-native-cached-image": "^1.2.3",
    "react-native-fs": "^2.1.0-rc.1",
    "crypto-js": "3.1.9-1",
    "lodash": "4.17.4",
    "url-parse": "1.1.7"
```

I even tried making my own implementation based off yours using AsyncStorage instead of hashing through SHA1 and ran into the same issue:

```//js
const _                         = require('lodash');
const UUID                      = require('uuid');
const RNFS                      = require('react-native-fs');
const { DocumentDirectoryPath } = RNFS;
const {
    AsyncStorage
} = require('react-native');

var activeDownloads = {};

export function checkExists(url) {
    return new Promise((resolve, reject) => {
        AsyncStorage.getItem(url, (error, result) => {
            if(_.isNull(error) && result) {
                const filePath = DocumentDirectoryPath + '/' + result;
                RNFS.exists(filePath).then(r => {
                    if(r) {
                        RNFS.stat(filePath).then((result) => {
                            if(!result.isFile()) return reject();
                            if(!result.size) return deleteFile(filePath).then(() => reject());
                            return resolve(filePath);
                        });
                    } else return reject();
                });
            } else {
                return reject();
            }
        });
    });
}

export function createPath(url) {
    const uuid = UUID.v4();
    return new Promise((resolve, reject) => {
        AsyncStorage.setItem(url, uuid, (error) => {
            if(_.isNull(error)) {
                const filePath = DocumentDirectoryPath + '/' + uuid;
                const parts = filePath.split('/');
                const dirPath = _.initial(parts).join('/');
                return RNFS.mkdir(dirPath, { NSURLIsExcludedFromBackupKey: true })
                    .then(r => resolve(filePath))
                    .catch(() => Promise.reject());
            }
        });
    });
}

export function deleteFile(toFile) {
    return RNFS.exists(filePath)
        .then((result) => result && RNFS.unlink)
        .catch((error) => {});
}

export function downloadImage(fromUrl, toFile) {
    if(!_.has(activeDownloads, toFile)) {
        activeDownloads[toFile] = new Promise((resolve, reject) => {
            RNFS.downloadFile({fromUrl, toFile}).promise
                .then((response) => {
                if(Math.floor(response.statusCode / 100) === 2) {
                    console.warn('DOWNLOADED OMFG');
                    resolve(toFile);
                } else {
                    return Promise.reject('Failed to download iamge');
                }
            })
            .catch((error) => {
                return deleteFile(toFile).then(() => reject(error));
            })
            .finally(() => {
                delete activeDownloads[toFile];
            });
        });
    }
    return activeDownloads[toFile];
}

export function isCacheable(url) {
    return _.isString(url) && (_.startsWith(url, 'http://') || _.startsWith(url, 'https://'));
}

export function createPrefetcer(list) {
    const urls = _.clone(list);
    return {
        next() {
            return urls.shift();
        }
    };
}

export function runPrefetchTask(prefetcher, options) {
    const url = prefetcher.next();
    if (!url) {
        return Promise.resolve();
    }
    if (isCacheable(url)) {
        return checkExists(url)
            .catch(() => {
                createPath(url)
                    .then((filePath) => downloadImage(url, filePath))
                    .then(() => runPrefetchTask(prefetcher, options));
            });
    }
    return runPrefetchTask(prefetcher);
}

export function cacheMultipleImages(urls) {
    const prefetcher = createPrefetcer(urls);
    const numberOfWorkers = urls.length;
    const promises = _.times(numberOfWorkers, () =>
        runPrefetchTask(prefetcher)
    );
    return Promise.all(promises);
}

const ReactNative = require('react-native');
const flattenStyle = ReactNative.StyleSheet.flatten;
const React = require('react');
const {
    View,
    Image,
    ActivityIndicator,
    Platform
} = require('react-native');

const styles = ReactNative.StyleSheet.create({
    image: { backgroundColor: 'transparent' },
    loader: { backgroundColor: 'transparent' },
    loaderPlaceholder: {
        backgroundColor: 'transparent',
        alignItems: 'center',
        justifyContent: 'center'
    }
});

export class CachedImage extends React.Component {

    constructor(props) {
        super(props);
        this._isMounted = false;
        this.state = {
            isCacheable: false,
            cachedImagePath: null };
    }

    safeSetState(newState) {
        if(!this._isMounted) return;
        return this.setState(newState);
    }

    componentWillMount() {
        this._isMounted = true;
        this.processSource(this.props.source);
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    componentWillReceiveProps(nextProps) {
        if (!_.isEqual(this.props.source, nextProps.source)) {
            this.processSource(nextProps.source);
        }
    }

    processSource(source) {
        const url = _.get(source, ['uri'], null);
        if (isCacheable(url)) {
            checkExists(url)
                .then((filePath) => {
                    this.safeSetState({
                        cachedImagePath: filePath
                    });
                })
                .catch(() => {
                    createPath(url)
                        .then((filePath) => downloadImage(url, filePath))
                        .then((filePath) => {
                        this.safeSetState({
                            cachedImagePath: filePath
                        });
                    })
                    .catch(err => {
                        this.safeSetState({
                            cachedImagePath: null
                        });
                    });
                this.safeSetState({
                    isCacheable: true
                });
            });
        } else {
            this.safeSetState({
                isCacheable: false
            });
        }
    }

    render() {
        if (this.state.isCacheable && !this.state.cachedImagePath) {
           return this.renderLoader();
       }

       const style = this.props.style || styles.image;
       const source = this.state.cachedImagePath ? {
               uri: 'file://' + this.state.cachedImagePath
           } : this.props.source;

           console.warn('Source:', source);
       return <Image
            ref='cachedImage'
           style={style}
           source={this.props.source} />;
    }

    renderLoader() {
        const imageStyle = [this.props.style, styles.loaderPlaceholder];

        const activityIndicatorProps = {};
        const activityIndicatorStyle = styles.loader;

        const source = this.props.defaultSource;

        // if the imageStyle has borderRadius it will break the loading image view on android
        // so we only show the ActivityIndicator
        if (Platform.OS === 'android' && flattenStyle(imageStyle).borderRadius) {
            return (
                <ActivityIndicator
                    {...activityIndicatorProps}
                    style={[imageStyle, activityIndicatorStyle]}/>
            );
        }
        // otherwise render an image with the defaultSource with the ActivityIndicator on top of it
        return (
            <Image
                source={source}
                style={imageStyle}>
                <ActivityIndicator
                    {...activityIndicatorProps}
                    style={activityIndicatorStyle}/>
            </Image>
        );
    }
}
```

**Comment by ChristianTucker** [comment]
Please note that while rendering the image, changing `{source}` to `source={this.props.source}` renders the correct image, albeit doesn't serve from disk. (Expected)

**Comment by ChristianTucker** [comment]
@kfiroo  I found the problem, it seems that the files downloaded over HTTPS are being stored without a suffix, IE: `.png` 

Navigating into the Filesystem of my device / simulator shows the images downloaded over HTTP as having a  `.jpg` or `.png` file extension, while everything downloaded over HTTPS does not. Appending `.png` or the appropriate file extension in the filesystem allows the image to load successfully. 

File path over HTTPS: user_photos%2Fc9bd1771-e08c-4fc4-986d-7658776e6491 
File path over HTTP: af2a25c75f3d6f3e249784bb83bd7873ba3c7c37.jpg

**Comment by kfiroo** [comment]
@ChristianTucker I have to say I have no idea what could go wrong here..
Let's try this really simplified version of downloading the file and trying to render an `Image` with the local file

```jsx
'use strict';

const _ = require('lodash');
const React = require('react');
const ReactNative = require('react-native');
const RNFS = require('react-native-fs');

const {
    View,
    Image,
    Dimensions,
    StyleSheet,
    AppRegistry,
    ActivityIndicator
} = ReactNative;

const {
    width
} = Dimensions.get('window');

const styles = StyleSheet.create({
    container: {
        flex: 1,
        marginTop: 20
    },
    image: {
        width,
        height: 300,
        backgroundColor: 'grey'
    }
});

const imageUrl = 'https://firebasestorage.googleapis.com/v0/b/qualify-college-dating.appspot.com/o/user_photos%2Fc9bd1771-e08c-4fc4-986d-7658776e6491?alt=media&token=9dae0c20-ef0a-40e9-ad87-25a8b8460f5c';
const cachedImagePath = RNFS.DocumentDirectoryPath + '/cache/test/image.jpg';

const CachedImageExample = React.createClass({

    getInitialState() {
        return {
            imageReady: false
        };
    },

    componentWillMount() {
        const downloadOptions = {
            fromUrl: imageUrl,
            toFile: cachedImagePath
        };
        const parts = cachedImagePath.split('/');
        const dirPath = _.initial(parts).join('/');
        RNFS.mkdir(dirPath, {NSURLIsExcludedFromBackupKey: true})
            .then(() => RNFS.downloadFile(downloadOptions).promise)
            .then(res => {
                if (Math.floor(res.statusCode / 100) == 2) {
                    this.setState({
                        imageReady: true
                    });
                } else {
                    throw new Error('Failed to successfully download image');
                }
            })
            .catch(err => {
                console.error(err);
            });
    },

    render() {
        const source = {
            uri: cachedImagePath
        };
        return (
            <View style={styles.container}>
                {this.state.imageReady ? (
                        <Image
                            source={source}
                            style={styles.image}
                        />
                    ) : (
                        <ActivityIndicator style={styles.image} />
                    )}
            </View>
        );
    }
});

AppRegistry.registerComponent('CachedImageExample', () => CachedImageExample);
```

**Comment by ChristianTucker** [comment]
@kfiroo  See my last comment (:

**Comment by kfiroo** [comment]
@ChristianTucker Oh! Great catch! Thanks!

Looks like the issue is with my code, not the `https`, take a look at this line https://github.com/kfiroo/react-native-cached-image/blob/master/ImageCacheProvider.js#L41

Your URL doesn't contain a file type and the code doesn't resolve the default type correctly, I will fix it and publish a new version soon.

Thank you :)

**Comment by ChristianTucker** [comment]
@kfiroo  Cool thanks.  Glad I didn't run into one of those delete and start over bugs.

**Comment by kfiroo** [comment]
@ChristianTucker one other thing, in your first report you had this snippet

```jsx
<CachedImage source={{uri: 'https://firebasestorage.googleapis.com/v0/b/qualify-college-dating.appspot.com/o/user_photos%2Fc9bd1771-e08c-4fc4-986d-7658776e6491?alt=media&token=9dae0c20-ef0a-40e9-ad87-25a8b8460f5c'}} style={{width: 300, height: 300}} />
```

Your URL has `%2F` in it (it's an encoded backslash `/`) is it intentional?
It will cause the the file name to be `user_photos%2Fc9bd1771-e08c-4fc4-986d-7658776e6491` and not simply `c9bd1771-e08c-4fc4-986d-7658776e6491` as it should, and might cause some issue later if you same the file names separately somewhere
Just a heads up :)

**Comment by ChristianTucker** [comment]
@kfiroo This is actually how the link is given to me after uploading from the CDN. I've never noticed it until you've pointed it out. The files are named with RFC4122 UUIDs so the Likelihood of a collision is extremely low. I'll go ahead and decode the value and run some maintenance on my database though, thanks for the catch.

**Comment by kfiroo** [comment]
@ChristianTucker published `v1.2.4`
Let me know if that works for you

**Comment by ChristianTucker** [comment]
@kfiroo  Resolved.
